File: 802. Find Eventual Safe States/eventualSafeNodes.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
    public:
        bool dfs(int pos, vector<int>& visit, vector<vector<int>>& grph){
            //base case
            if(visit[pos] == 2)return true;
    
            visit[pos] = 1;
    
            for(auto item : grph[pos]){
                if(visit[item] == 1 || !dfs(item, visit, grph)){
                    return false;
                }
            }
    
            visit[pos] = 2;
    
            return true;
        }
        vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
            int sz = graph.size();
            vector<int> visit(sz, 0);
            vector<int> ans;
    
    
            for(int i=0 ; i<sz; i++){
                if(dfs(i,visit, graph))ans.push_back(i);
            }
    
            return ans;
    
        }
    };

File: 802. Find Eventual Safe States/eventualSafeNodes.js
/**
 * @param {number[][]} graph
 * @return {number[]}
 */
var eventualSafeNodes = function (graph) {
    let nodes = graph.length
    let ans = []
    let visit = new Array(nodes).fill(0);
    let stack = []


    function dfs(node) {
        if (visit[node] === 2) return true

        visit[node] = 1;

        for (let item of graph[node]) {
            if (visit[item] === 1 || !dfs(item)) {
                return false
            }
        }

        visit[node] = 2
        return true;
    }


    for (let i = 0; i < nodes; i++) {
        if (dfs(i)) ans.push(i)
    }

    return ans;
};

/**
- state for visit marker
=> 0 = unvisited
=> 1 = visited
=> 2 = safe area
 */